fmn/api/handlers/misc.py

```diff
@@ -37,39 +38,50 @@ def get_applications():
     return applications
 
 
-@router.get("/artifacts/owned", response_model=list[dict[str, str]], tags=["misc"])
-async def get_owned_artifacts(
+@router.get("/artifacts", response_model=list[api_models.Artifact], tags=["misc"])
+async def get_artifacts(
+    names: list[str] = Query(default=[]),
     users: list[str] = Query(default=[]),
     groups: list[str] = Query(default=[]),
     distgit_proxy: PagureAsyncProxy = Depends(get_distgit_proxy),
 ):
+    """This handler queries artifacts from Pagure
+
+    Proxying Pagure queries lets the API cache results to reduce load on the
+    backend service.
+
+    :param names: Name patterns of artifacts which should be returned
+
+    :param users: Names of users whose artifacts should be returned
+
+    :param groups: Names of groups whose artifacts should be returned
+    """
     artifacts = {}
 
+    for name_pattern in names:
+        artifacts.update(
+            (project["fullname"], {"type": project["namespace"], "name": project["name"]})
+            for project in await distgit_proxy.get_projects(pattern=name_pattern)
+            if ArtifactType.has_value(project["namespace"])
+        )
+
     for user in users:
-        for project in await distgit_proxy.get_user_projects(username=user):
-            artifacts[project["fullname"]] = {"type": project["namespace"], "name": project["name"]}
+        artifacts.update(
+            (project["fullname"], {"type": project["namespace"], "name": project["name"]})
+            for project in await distgit_proxy.get_user_projects(username=user)
+            if ArtifactType.has_value(project["namespace"])
+        )
 
     for group in groups:
-        for project in await distgit_proxy.get_group_projects(name=group):
-            artifacts[project["fullname"]] = {"type": project["namespace"], "name": project["name"]}
+        artifacts.update(
+            (project["fullname"], {"type": project["namespace"], "name": project["name"]})
+            for project in await distgit_proxy.get_group_projects(name=group)
+            if ArtifactType.has_value(project["namespace"])
+        )
 
     return sorted(artifacts.values(), key=lambda a: (a["name"], a["type"]))
```

Comment:
It looks like you are using a dict for de-duplication, because the values are dicts, and therefore not hashable. How about something like this:
```python
artifacts = set()
for ...:
    artifacts.add(
        (project["name"], project["namespace"])
        for project in await distgit_proxy....
        if ArtifactType.has_value(project["namespace"])
    )
return [{"name": a[0], "type": a[1]} for a in sorted(artifacts)]
```

Reply:
This together with the suggestion above should make the handler pretty slim, I like it.